### ISSUE TITLE
`compute_metrics` show better results than `generate` because target data leaks

### ISSUE THREAD

**Issue by pperle** [huggingface/transformers#10245]
## Environment info

- `transformers` version: 4.3.2
- Platform: Linux-5.8.18-050818-generic-x86_64-with-glibc2.10
- Python version: 3.8.5
- PyTorch version (GPU?): 1.7.1 (True)
- Tensorflow version (GPU?): not installed (NA)
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

## Information

Model I am using:
- t5

The problem arises when using:
- my own modified scripts: (give details below)

The tasks I am working on is:
- my own task or dataset: (give details below)

## To reproduce

1. train model with `compute_metrics` function to monitor metrics
2. use `generate` to predict text with trained model

## Expected behavior

I expect the metrics of `compute_metrics` to be equal to my generated text.

## More information

While training, I used `compute_metrics` to calculate the metric on my validation set every X steps. I was surprised to see that after training my model did not perform as expected using the `generate` function provided by huggingface.

After some digging through the code I think I understand what the problem is. [`compute_metrics`](https://github.com/huggingface/transformers/blob/cd48078ce59a195473729759c76d88ae612b0f7a/src/transformers/trainer.py#L1665) takes as input `preds`, which is [a collection of `logits`](https://github.com/huggingface/transformers/blob/cd48078ce59a195473729759c76d88ae612b0f7a/src/transformers/trainer.py#L1635) from [`prediction_step`](https://github.com/huggingface/transformers/blob/cd48078ce59a195473729759c76d88ae612b0f7a/src/transformers/trainer.py#L1630) which internally [calls `model` with the inputs and targets of the model](https://github.com/huggingface/transformers/blob/cd48078ce59a195473729759c76d88ae612b0f7a/src/transformers/trainer.py#L1733).
This means that the target text leaks into `preds.predictions` because `mode.forward` used the targets as [input for the decoder](https://github.com/huggingface/transformers/blob/1cd16512dc8060aa8c2419664f9cb83813ade4d5/src/transformers/models/t5/modeling_t5.py#L1331). This makes the metrics of `compute_metrics` seem much better than they really are.

In my opinion the target data should not be used to create `preds.predictions`. Maybe the `generate` function is a better fit.

**Comment by LysandreJik** [comment]
Pinging @patrickvonplaten and @sgugger

**Comment by sgugger** [comment]
Did you use the flag `--predict_with_generate`? It's there just for this: predicting using the `generate` method and the labels are then not passed (except to compute the loss).

**Comment by pperle** [comment]
Thank you for the hint. I followed this tutorial [this example](https://huggingface.co/transformers/custom_datasets.html#fine-tuning-with-trainer) and used [`TrainingArguments`](https://huggingface.co/transformers/main_classes/trainer.html#trainingarguments), which to not have the `predict_with_generate` option, instead of [`Seq2SeqTrainingArguments`](https://huggingface.co/transformers/main_classes/trainer.html#transformers.Seq2SeqTrainingArguments).

Maybe it's just me, but I think the `predict_with_generate` option should be described more visibly. I found it only in [`Seq2SeqTrainingArguments`](https://huggingface.co/transformers/main_classes/trainer.html#transformers.Seq2SeqTrainingArguments) and in none of the examples. Also, none of the examples in the documentation use `Seq2SeqTrainingArguments`.

If you don't think the documentation should be updated you can close this issue since my confusion has been resolved. Thank you.

**Comment by sgugger** [comment]
Yes the documentation is missing a seq2seq example. This is because we have been working on the design recently. The most up-to-date example you should use as reference is the [run_seq2seq script](https://github.com/huggingface/transformers/blob/master/examples/seq2seq/run_seq2seq.py).
